### PR TITLE
Support custom region tags in samples

### PR DESCRIPTION
**Pull request by vchudnov-g**
- The default region tag is "sample"
- Custom region tags can be specified in the GAPIC YAML. The region tag specification accepts (but does not require) `%m`, `%c`, and `%v` placeholders for method name, calling form name, and value set id.
- If the sample specification would result in differently configured samples being written to the same path, an exception is thrown. This can happen if the same value set and calling form combo is specified more than once (eg with wildcards) with different region names.

### CODE REVIEW

File: src/main/java/com/google/api/codegen/transformer/SampleTransformer.java [hunk 1]
```diff
@@ -195,37 +197,67 @@ public void generateSamples(
...
       }
     }
     return methodSampleViews;
   }
 
+  /**
+   * Creates a region tag from the spec, replacing any {@code %m}, {@code %c}, and {@code %v}
+   * placeholders with the method name, calling form name, and value set id, respectively.
+   */
+  private String regionTagFromSpec(
```
Comment (hzyi-google) [inline src/main/java/com/google/api/codegen/transformer/SampleTransformer.java]: this method could be declared static?

File: src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * {@code SampleFileRegistry} is used to verify that the samples we generate with different
+ * parameters have different paths, so they don't clobber each other.
+ */
+public class SampleFileRegistry {
+
+  Map<String, SampleInfo> files = new HashMap<>();
```
Comment (hzyi-google) [inline src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java]: Can you make this `private final`?

File: src/main/java/com/google/api/codegen/config/SampleSpec.java [hunk 3]
```diff
@@ -46,16 +51,49 @@
...
+    }
+  }
+
   public SampleSpec(MethodConfigProto methodConfigProto) {
     specified = methodConfigProto.hasSamples();
     sampleConfiguration = methodConfigProto.getSamples();
     valueSets = methodConfigProto.getSampleValueSetsList();
 
-    HashSet<String> ids = new HashSet<>();
+    HashMap<String, SampleValueSet> setMap = new HashMap<>();
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/config/SampleSpec.java]: I don't think this loop has to change at all. We're not actually using the `valueSet` themselves for anything here right?

File: src/main/java/com/google/api/codegen/config/SampleSpec.java [hunk 4]
```diff
@@ -101,21 +134,26 @@ private static boolean expressionMatchesId(String expression, String id) {
...
-        .stream()
-        .filter(
-            valueSet ->
-                valueSetExpressions
-                    .stream()
-                    .anyMatch(expression -> expressionMatchesId(expression, valueSet.getId())))
-        .collect(ImmutableList.toImmutableList());
+    // Construct a `ValueSetAndTags` for each sample specified in each element of `matchingSamples`.
+    List<ValueSetAndTags> result = new ArrayList<>();
+    Set<String> generated = new HashSet<>();
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/config/SampleSpec.java]: This seems unused?

File: src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+              .method(method)
+              .callingForm(callingForm)
+              .valueSet(valueSet)
+              .regionTag(regionTag)
+              .build());
+    } else if (!(previous.path().equals(path)
+        && previous.method().equals(method)
+        && previous.callingForm().equals(callingForm)
+        && previous.valueSet().equals(valueSet)
+        && previous.regionTag().equals(regionTag))) {
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java]: Just use `this.equals(previous)`, `@AutoValue` already makes one for us.

File: src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      throw new IllegalArgumentException(
+          String.format(
+              "conflicting configurations for sample \"%s\":\n"
+                  + "%s\n%s\n"
+                  + "  {method:\"%s\", calling form:\"%s\", value set:\"%s\", region tag:\"%s\"}\n"
+                  + "  {method:\"%s\", calling form:\"%s\", value set:\"%s\", region tag:\"%s\"}\n",
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java]: Just use `toString()`, AutoValue already implements it.

File: src/main/java/com/google/api/codegen/transformer/SampleTransformer.java [hunk 7]
```diff
@@ -195,37 +197,67 @@ public void generateSamples(
...
+      String spec, String methodName, CallingForm callingForm, String valueSetId) {
+    final String DEFAULT_REGION_SPEC = "sample";
+
+    if (spec == null || spec.isEmpty()) {
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/SampleTransformer.java]: Prefer Guava's `isNullOrEmpty`

File: src/main/java/com/google/api/codegen/transformer/SampleTransformer.java (hunk 1 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/transformer/SampleTransformer.java]: done

File: src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java (hunk 2 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java]: done

File: src/main/java/com/google/api/codegen/transformer/SampleTransformer.java (hunk 7 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/transformer/SampleTransformer.java]: Used `Strings.isNullOrEmpty`

File: src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java (hunk 6 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java]: OK, but at the price of always building a new `SampleInfo`

File: src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java (hunk 5 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java]: OK, but at the price of always building a new `SampleInfo`

File: src/main/java/com/google/api/codegen/config/SampleSpec.java (hunk 4 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/config/SampleSpec.java]: Good catch! This was cruft from when I was checking for uniqueness here.

File: src/main/java/com/google/api/codegen/config/SampleSpec.java (hunk 3 above)
Comment (vchudnov-g) [reply src/main/java/com/google/api/codegen/config/SampleSpec.java]: Good catch! Cruft from an earlier approach

**Comment by vchudnov-g** [comment]
PTAL

**Pull request merged**